.github/workflows/ci.yml

```diff
@@ -48,6 +47,13 @@ jobs:
   cargo-test:
     strategy:
       matrix:
+        os: [ubuntu, windows, macos]
+        py:
+          - "3.8"
+          - "3.9"
+          - "3.10"
+          - "3.11"
+          - "3.12"
```

Comment:
All these test should run on a single runner in our case since building uv (as a rust project) and running tests with a lot of integration test is much more expensive than for virtualenv, a small scoped python project.

Reply:
😭 makes sense - i'll look into that maybe later today. thanks for pointing it out early

Reply:
I believe that the Python version here is only used to bootstrap Python itself. We then have code that downloads specific pinned Python versions for the test suites.

Reply:
just to clarify - so the preferred solution here is to set up all of the python versions on a single runner (for each os) and vary the versions _only_ for the venv test? it does seem more common to use a runner per each runtime version in other projects i'm familiar with, but also i agree that it's less wasteful with single runner for all versions if we are varying the version only for the venv tests.

Reply:
We had problems where e.g. python-build-standalone would fail when creating a venv from another venv, that the paths of debian/ubuntu pythons were different than those elsewhere, windows store python failing on canonicalize, pypy has a different venv structure, etc. It would be great if we could test as many of these scenarios on CI to avoid regressions, and because we can't install them automatically for our local test suite (e.g. the `apt install python` isn't available on fedora and we can't modify a developers system by adding extra pythons to the registry). How exactly to best implement this - i'm not sure! Maybe something like installing all pythons and passing them to the venv tests as an env var, or maybe we have to run the venv tests multiple times, this requires some experimentation.

For the majority of our integration tests, we run with python 3.12 from python-build-standalone from the bootstrap script, some use different versions to check we handle markers, builds and caching correctly. We don't want to them multiple times since they are slow (lots of network, decompression and io) and the snapshots expect python 3.12.

Reply:
Ah the joys of supporting multiple platforms 😆 - I'll try to take a look this coming weekend, thanks for the detailed context!